search-replace_test.go

```diff
@@ -93,11 +163,59 @@ func TestReplace(t *testing.T) {
 			in:  []byte(`s:17:\"https://spock.com\";`),
 			out: []byte(`s:15:\"http://🖖.com\";`),
 		},
+		{
+			testName: "search and replace with different lengths",
+
+			from: []byte("hello"),
+			to:   []byte("goodbye"),
+
+			in:  []byte(`s:11:\"hello-world\";`),
+			out: []byte(`s:13:\"goodbye-world\";`),
+		},
+		{
+			testName: "serialized CSS",
+			from:     []byte(`https://uss-enterprise.com`),
+			to:       []byte(`https://ncc-1701-d.space`),
+			in:       []byte(`a:2:{s:3:\"key\";s:5:\"value\";s:3:\"css\";s:208:\"body { color: #123456;\r\nborder-bottom: none; }\r\ndiv.bg { background: url('https://uss-enterprise.com/wp-content/uploads/main-bg.gif');\r\n  background-position: left center;\r\n    background-repeat: no-repeat; }\";}`),
+			out:      []byte(`a:2:{s:3:\"key\";s:5:\"value\";s:3:\"css\";s:206:\"body { color: #123456;\r\nborder-bottom: none; }\r\ndiv.bg { background: url('https://ncc-1701-d.space/wp-content/uploads/main-bg.gif');\r\n  background-position: left center;\r\n    background-repeat: no-repeat; }\";}`),
+		},
+		{
+			testName: "string encoded by both MySQL and PHP",
+
+			from: []byte(`http:\\/\\/example\\.com`),
+			to:   []byte(`http:\\/\\/example2\\.com`),
+			in:   []byte(`s:37:\"\\s=\\shttp_get\\(\'http:\\/\\/example\\.com\";`),
+			out:  []byte(`s:38:\"\\s=\\shttp_get\\(\'http:\\/\\/example2\\.com\";`),
+		},
+		{
+			testName: "non-serial replacement trying to apply itself to serial replacement",
+
+			from: []byte(`example`),
+			to:   []byte(`example2`),
+			in:   []byte(`('example'),('s:37:\"\\s=\\shttp_get\\(\'http:\\/\\/example\\.com\";')`),
+			out:  []byte(`('example2'),('s:38:\"\\s=\\shttp_get\\(\'http:\\/\\/example2\\.com\";')`),
+		},
+		{
+			testName: "lots of encoding",
+			from:     []byte(`\\c\\d\\e`),
+			to:       []byte(`\\x`),
+			in:       []byte(`s:18:\"\\a\\b\\c\\d\\e\\f\\g\\h\";\";`),
+			out:      []byte(`s:14:\"\\a\\b\\x\\f\\g\\h\";\";`),
+		},
+		{
+			testName: "search and replace with different lengths",
+
+			from: []byte("bbbbbbbbbb"),
+			to:   []byte("ccccccccccccccc"),
+
+			in:  []byte(`s:20:\"aaaaabbbbbbbbbbaaaaa\";`),
+			out: []byte(`s:25:\"aaaaacccccccccccccccaaaaa\";`),
+		},
```

Comment:
I've found a few edge cases on the php-src tests: https://github.com/php/php-src/tree/master/ext/standard/tests/serialize

Not sure if it's worth fixing them all.

```suggestion
		},
		{
			testName: "exotic letters",

			from: []byte("åäöÅÄÖ"),
			to:   []byte("---åäöÅÄÖ---"),

			in:  []byte(`a:1:{s:12:"åäöÅÄÖ";s:12:"åäöÅÄÖ";}`),
			out: []byte(`a:1:{s:18:"---åäöÅÄÖ---";s:18:"---åäöÅÄÖ---";}`),
		},
		{
			testName: "complex array",

			from: []byte("hallo"),
			to:   []byte("world!"),

			in:  []byte(`a:6:{i:0;i:1;i:1;d:1.1;i:2;s:5:"hallo";i:3;N;i:4;b:1;i:5;a:0:{}}`),
			out: []byte(`a:6:{i:0;i:1;i:1;d:1.1;i:2;s:6:"world!";i:3;N;i:4;b:1;i:5;a:0:{}}`),
		},
		{
			testName: "object",

			from: []byte("hello"),
			to:   []byte("world!"),

			in:  []byte(`O:8:"stdClass":1:{s:3:"key";s:5:"hello";}`),
			out: []byte(`O:8:"stdClass":1:{s:3:"key";s:6:"world!";}`),
		},
		{
			testName: "object with recursion",

			from: []byte("hello!"),
			to:   []byte("world!!!"),

			in:  []byte(`O:8:"stdClass":2:{s:4:"obj2";O:8:"stdClass":1:{s:4:"obj3";O:8:"stdClass":1:{s:4:"obj1";r:1;}}s:5:"hello";s:6:"hello!";}`),
			out: []byte(`O:8:"stdClass":2:{s:4:"obj2";O:8:"stdClass":1:{s:4:"obj3";O:8:"stdClass":1:{s:4:"obj1";r:1;}}s:5:"world";s:8:"world!!!";}`),
		},
		{
			testName: "replace object key with invalid value",

			from: []byte("valid_key"),
			to:   []byte("invalid_key!!"),

			in:  []byte(`O:8:"stdClass":1:{s:9:"valid_key";s:4:"data";}`),
			out: []byte(`O:8:"stdClass":1:{s:9:"valid_key";s:4:"data";}`), //Should we replace the key or not?
		},
		{
			testName: "object with replace in key and value",

			from: []byte("t1"),
			to:   []byte("t999"),

			in:  []byte(`O:8:"stdClass":1:{s:2:"t1";a:1:{s:2:"t1";a:1:{i:0;s:2:"t1";}}}`),
			out: []byte(`O:8:"stdClass":1:{s:4:"t999";a:1:{s:4:"t999";a:1:{i:0;s:4:"t999";}}}`),
		},		
```

Reply:
Not really worth it. I pondered over it but ultimately decided not to. We're inherently using bytes and bytelength so the encoding doesn't really matter.

a 2-byte utf-8 character, when encoded to ascii only for example... is still 2 bytes - just that it's now 2 characters. We already have tests for emojis so that's a suitable equivalent for accented letters.

Reply:
As for the object keys... probably not worth it, and we might want to replace the keys as well, too. Think of key-value where the key is domain, and value is... something else.

Reply:
I'll double check these tests though - as I'm sure the ones with accented characters should work.

Reply:
Okay, yes, worth fixing. Thanks for these @luismulinari 

Kinda just assumed that it's working

Reply:
Just double-checked the tests - You need to escape it the same way as mysql escapes it i.e. put it in a db, and make a mysqldump.

The search-replace serializer does not apply to non-sql-dumps to keep the complexity low. We might want to consider it later once we've seen usages related to it.

Reply:
i.e. instead of this:
```golang
		{
			testName: "exotic letters",
			from:     []byte("åäöÅÄÖ"),
			to:       []byte("---åäöÅÄÖ---"),
			in:       []byte(`a:1:{s:12:"åäöÅÄÖ";s:12:"åäöÅÄÖ";}`),
			out:      []byte(`a:1:{s:18:"---åäöÅÄÖ---";s:18:"---åäöÅÄÖ---";}`),
		},
```


Do this (notice the escaped quotes)

```golang
		{
			testName: "exotic letters",
			from:     []byte("åäöÅÄÖ"),
			to:       []byte("---åäöÅÄÖ---"),
			in:       []byte(`a:1:{s:12:\"åäöÅÄÖ\";s:12:\"åäöÅÄÖ\";}`),
			out:      []byte(`a:1:{s:18:\"---åäöÅÄÖ---\";s:18:\"---åäöÅÄÖ---\";}`),
		},
```

Reply:
```suggestion
		},
		{
			testName: "exotic letters",
			from:     []byte("åäöÅÄÖ"),
			to:       []byte("---åäöÅÄÖ---"),
			in:       []byte(`a:1:{s:12:\"åäöÅÄÖ\";s:12:\"åäöÅÄÖ\";}`),
			out:      []byte(`a:1:{s:18:\"---åäöÅÄÖ---\";s:18:\"---åäöÅÄÖ---\";}`),
		},
		{
			testName: "complex array",
			from:     []byte("hallo"),
			to:       []byte("world!"),
			in:       []byte(`a:6:{i:0;i:1;i:1;d:1.1;i:2;s:5:\"hallo\";i:3;N;i:4;b:1;i:5;a:0:{}}`),
			out:      []byte(`a:6:{i:0;i:1;i:1;d:1.1;i:2;s:6:\"world!\";i:3;N;i:4;b:1;i:5;a:0:{}}`),
		},
		{
			testName: "object",
			from:     []byte("hello"),
			to:       []byte("world!"),
			in:       []byte(`O:8:\"stdClass\":1:{s:3:\"key\";s:5:\"hello\";}`),
			out:      []byte(`O:8:\"stdClass\":1:{s:3:\"key\";s:6:\"world!\";}`),
		},
		{
			testName: "object with recursion",
			from:     []byte("hello!"),
			to:       []byte("world!!!"),
			in:       []byte(`O:8:\"stdClass\":2:{s:4:\"obj2\";O:8:\"stdClass\":1:{s:4:\"obj3\";O:8:\"stdClass\":1:{s:4:\"obj1\";r:1;}}s:5:\"hello\";s:6:\"hello!\";}`),
			out:      []byte(`O:8:\"stdClass\":2:{s:4:\"obj2\";O:8:\"stdClass\":1:{s:4:\"obj3\";O:8:\"stdClass\":1:{s:4:\"obj1\";r:1;}}s:5:\"hello\";s:8:\"world!!!\";}`),
		},
		{
			testName: "replace object key with invalid value",
			from:     []byte("valid_key"),
			to:       []byte("invalid_key!!"),
			in:       []byte(`O:8:\"stdClass\":1:{s:9:\"valid_key\";s:4:\"data\";}`),
			out:      []byte(`O:8:\"stdClass\":1:{s:13:\"invalid_key!!\";s:4:\"data\";}`), //Should we replace the key or not?
		},
		{
			testName: "object with replace in key and value",
			from:     []byte("t1"),
			to:       []byte("t999"),
			in:       []byte(`O:8:\"stdClass\":1:{s:2:\"t1\";a:1:{s:2:\"t1\";a:1:{i:0;s:2:\"t1\";}}}`),
			out:      []byte(`O:8:\"stdClass\":1:{s:4:\"t999\";a:1:{s:4:\"t999\";a:1:{i:0;s:4:\"t999\";}}}`),
		},
```